domain/state.go

```diff
@@ -99,6 +99,41 @@ func (st *StateBase) Prepare(query string, typeSamples ...any) (*sqlair.Statemen
 	return stmt, nil
 }
 
+// RunTx executes the closure function within the scope of a transaction.
+// The closure is passed a TxContext that can be passed on to state
+// functions, so that they can perform work within that same transaction.
+// The closure will be retried according to the transaction retry semantics,
+// if the transaction fails due to transient errors.
+// The closure should only be used to perform state changes and must not be used
+// to execute queries outside of the state scope.
+func (st *StateBase) RunTx(ctx context.Context, fn func(TxContext) error) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Annotate(err, "getting database")
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// The txContext is created with the transaction and passed to the
+		// closure function. This ensures that the transaction is always
+		// available to the closure. Once the transaction is complete, the
+		// transaction is removed from the txContext. This is to prevent
+		// the transaction from being used outside of the transaction scope.
+		// This will prevent any references to the sqlair.TX from being held
+		// outside of the transaction scope.
+
+		// TODO (stickupkid): The txContext can be pooled on the StateBase to
+		// reduce the number of allocations. Attempting to push the tx into
+		// the context would prevent that as a viable option.
+		txCtx := &txContext{
+			Context: ctx,
+			t:       tx,
+		}
+		defer txCtx.removeTx()
```

Comment:
I guess the question becomes, should we go to this effort? I'm currently split on this, for the simple reason that the State interface for some services will expose the following:

```go
type State interface {
	GetFoo(domain.TxContext) error
}
```

Before this change, the idea of a transaction in the service layer was not a concept. The contract of a State interface will now expose this, so it's plausible to capture the TxContext outside of the RunTx closure and attempt to run it. It has meaning now, it is useful. It's not enforceable by static analysis, they're both intended to be used in the Service package. Using sqlair.TX or sql.Tx inside of a service package was enforceable through tooling.

I am happy to reconsider, just be aware that the training wheels are off here, and no matter how much we attempt to capture it in a code review, review fatigue is real.

Reply:
Keep, it. Just call it `Close` once we rename.